feat: refactor battle move flyout, add support for mid-battle moveset changes

Bertie690 opened the pull request:
## What are the changes the user will see?
The battle move flyout now tracks PP ups correctly, as well as mid-battle changes to movesets.
This includes:
 - Sketch
 - Leveling up and replacing a move (currently useless since we don't track player moves used in the flyout, but in theory could be used for a revamped enemy AI)
 - Mimic
 - Transform/Imposter

The last 2 of the above effects (due to being temporary) will reset once the target's `SummonData` is cleared (i.e. once it switches out or faints).

Furthermore, if a Pokemon transforms into an opposing Pokemon, the enemy's current flyout info will be revealed to the player, similar to how Showdown will reveal the moveset of any opponent you successfully Transform into.

Fixes #5150

## Why am I making these changes?
Not only is this nice QoL and fixes a longstanding bug, the information added by this PR could (in theory) be used by a prospective trainer AI or other system to better gauge the revealed moves of its opponent.

The relevant information would almost certainly be stored somewhere else, but the new event listeners can be hooked into relatively easily if the need arises.

## What are the changes from a developer perspective?

### `BattleFlyout`

1. A new `BattleSceneEvent`, `MovesetChangedEvent` has been added.
This takes a `PokemonMove` (instead of a generic `Move` instance), and is emitted whenever a `Pokemon` has its moveset changed in any way (both from effects like Copycat/etc and PP consumption/restoration). Depending on whether the attached `PokemonMove` is found in its normal or temporary movesets, the info will be added to the normal info or temporary info arrays respectively.
   - This serves as a drop-in replacement for `BerryUsedEvent` (which was _only_ used to track Leppa Berry PP restoration) and `MoveUsedEvent` (which was effectively just a "consume PP" event with a really misleading name).
2. The `BattleFlyout`'s move info array has been split in 2, now consisting of 2 parallel sparse arrays containing data for all moves having been used by or added to the attached Pokemon's moveset. The former tracks the Pokemon's regular moves (Sketch, etc.), while the latter is used for moves found exclusively in the temporary `SummonData` moveset (such as by Transform/Mimic).
3. Added a new event emitter to clear a Pokemon's battle flyout's temporary moveset information when its summon data is cleared. This ensures that temporary move info from Transform and Mimic will be retained only until their effects wear off.
4. Added code to `PokemonTransformPhase` to copy move flyout data upon transforming into an opponent.

## Screenshots/Videos

https://github.com/user-attachments/assets/a1627c0c-4624-4131-aa43-94c44fa9dbea

## How to test the changes?
Use local overrides and verify all of the following work:
1. Normal moves update the flyout and are kept after switching
2. Enemies transforming into player pokemon updates the flyout with the player's current moveset
3. Mimic updates the flyout with the learned move temporarily until leaving the field
4. Sketch updates the _permanent_ flyout

2, 3 and 4 would be easiest to test by using Whirlwind or similar to force the enemy out in a 2 mon trainer battle.

I have not had the time to test this recently since last beta merge, so any help would be greatly recommended

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
- [x] Are all unit tests still passing? (`pnpm test:silent`)
- [x] Have I provided screenshots/videos of the changes (if applicable)?

Bertie690 committed:
Started fixing the battle scene flyout

Bertie690 committed:
Removed unused `turnInitEvent

Bertie690 committed:
Continued fixing up the battle show thingy

Bertie690 committed:
Renamed some battle flyout stuff

Bertie690 committed:
Fixed battler thingy

Bertie690 committed:
Re-added typed event target

Bertie690 committed:
Merge branch 'beta' into battle-move-flyout

Bertie690 committed:
Merge remote-tracking branch 'upstream/beta' into battle-move-flyout

Bertie690 committed:
Fixed missing imports

Bertie690 committed:
Merge remote-tracking branch 'upstream/beta' into battle-move-flyout

Bertie690 committed:
Ran Biome

Bertie690 committed:
Merge remote-tracking branch 'upstream/beta' into battle-move-flyout

Bertie690 committed:
Merge remote-tracking branch 'upstream/beta' into battle-move-flyout

# Conflicts:
#	src/phases/move-phase.ts

Bertie690 committed:
Merge branch 'beta' into battle-move-flyout

Bertie690 committed:
Merge remote-tracking branch 'upstream/beta' into battle-move-flyout

Bertie690 committed:
Merge remote-tracking branch 'upstream/beta' into battle-move-flyout

Bertie690 committed:
Fix some errors

Bertie690 committed:
Merge branch 'beta' into battle-move-flyout

Bertie690 committed:
Merge remote-tracking branch 'upstream/beta' into battle-move-flyout

Bertie690 committed:
fix stuff

Bertie690 committed:
copy over new typed event target code

SirzBenjie reviewed (dismissed):
Events for UI seem to be a really good pattern.

DayKev committed:
Apply review

- Remove obsolete `null` check in `pokemon-transform-phase.ts`

- Apply `readonly` to various fields in `BattleFlyout`

- Rename `BattleFlyout#player` to `isPlayer`

- Minor changes to docs/comments, including adding missing type imports

DayKev committed:
Merge branch 'beta' into battle-move-flyout

DayKev committed:
Fix merge issue

DayKev committed:
Merge branch 'beta' into battle-move-flyout

Bertie690 committed:
remove redundant field

Bertie690 committed:
remove useless params from class

Bertie690 committed:
Merge remote-tracking branch 'upstream/beta' into battle-move-flyout

Bertie690 committed:
Ran Biome

Bertie690 committed:
fix background drawing issues

fabske0 commented:
The flyout doesn't appear at all after the last commit.
<details><summary>Image</summary>

<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/3b3d7f56-b2cd-4294-8115-918ed3d17f1e" />

</details>

Bertie690 commented:
AFHFEJEJFHFHENEJ

Bertie690 committed:
remove `flyoutParent` + fix background layering issues

Bertie690 committed:
Merge remote-tracking branch 'upstream/beta' into battle-move-flyout

SirzBenjie committed:
Merge branch 'beta' into battle-move-flyout

Bertie690 committed:
Merge branch 'beta' into battle-move-flyout

Bertie690 committed:
fix merge errors

Bertie690 committed:
run bioe
